dpnp/linalg/dpnp_utils_linalg.py

```diff
@@ -267,101 +267,76 @@ def _batched_solve(a, b, exec_q, res_usm_type, res_type):
+        sycl_queue=exec_q,
+    )
+    ht_list_ev.append(a_ht_copy_ev)
 
-        # oneMKL LAPACK assumes fortran-like array as input, so
-        # allocate a memory with 'F' order for dpnp array of multiple
-        # dependent variables array
-        val_vecs[i] = dpnp.empty_like(
-            b[i], order="F", dtype=res_type, usm_type=res_usm_type
-        )
+    # oneMKL LAPACK gesv overwrites `b` and assumes fortran-like array
+    # as input.
+    b_f = dpnp.empty_like(b, order="F", dtype=res_type, usm_type=res_usm_type)
 
-        # use DPCTL tensor function to fill the array of multiple dependent
-        # variables with content from the input arrays
-        b_ht_copy_ev[i], b_copy_ev = ti._copy_usm_ndarray_into_usm_ndarray(
-            src=b_usm_arr[i],
-            dst=val_vecs[i].get_array(),
-            sycl_queue=b.sycl_queue,
-        )
+    b_ht_copy_ev, b_copy_ev = ti._copy_usm_ndarray_into_usm_ndarray(
+        src=b_usm_arr,
+        dst=b_f.get_array(),
+        sycl_queue=exec_q,
+    )
+    ht_list_ev.append(b_ht_copy_ev)
 
-        # Call the LAPACK extension function _gesv to solve the system of
-        # linear equations using a portion of the coefficient square matrix
-        # and a corresponding portion of the dependent variables array.
-        ht_lapack_ev[i], _ = li._gesv(
-            exec_q,
-            coeff_vecs[i].get_array(),
-            val_vecs[i].get_array(),
-            depends=[a_copy_ev, b_copy_ev],
-        )
+    ht_lapack_ev, _ = li._gesv_batch(
+        exec_q,
+        a_f.get_array(),
+        b_f.get_array(),
+        depends=[a_copy_ev, b_copy_ev],
+    )
 
-        # TODO: Remove this w/a when MKLD-17201 is solved.
-        # Waiting for a host task executing an OneMKL LAPACK gesv call
-        # on CPU causes deadlock due to serialization of all host tasks
-        # in the queue.
-        # We need to wait for each host tasks before calling _gesv to avoid
-        # deadlock.
-        if is_cpu_device:
-            ht_lapack_ev[i].wait()
-            b_ht_copy_ev[i].wait()
+    ht_list_ev.append(ht_lapack_ev)
 
-    for i in range(batch_size):
-        ht_lapack_ev[i].wait()
-        b_ht_copy_ev[i].wait()
-        a_ht_copy_ev[i].wait()
+    dpctl.SyclEvent.wait_for(ht_list_ev)
 
-    # combine the list of solutions into a single array
-    out_v = dpnp.array(
-        val_vecs, order=b_order, dtype=res_type, usm_type=res_usm_type
-    )
-    if reshape:
-        # shape of the out_v must be equal to the shape of the array of
-        # dependent variables
-        out_v = out_v.reshape(orig_shape_b)
-    return out_v
+    # Gesv call overwtires `b` in Fortran order, reorder the axes
+    # to match C order by moving the last axis to the front and
+    # reshape it back to the original shape of `b`.
+    v = dpnp.moveaxis(b_f, -1, 0).reshape(b_shape)
+
+    # dpnp.moveaxis can make the array non-contiguous if it is not 2D
+    if b.ndim > 2:
+        v = dpnp.ascontiguousarray(v)
```

Comment:
Is that only to be aligned with numpy? Do we need to add an explicit comment here? Otherwise the motivation isn't fully clear.

Reply:
I updated this comment in the last commit